Add _foreach_max

janeyx99 opened the pull request:
This PR adds _foreach_max support, the second reduction foreach op we have :D 

I did have to change the autogen slightly for foreach. I can promise that the existing foreach ops' derivative behavior has not changed as I've added a skip list for the harder requirement I am setting (that the arg list should match in length). I needed to add this requirement as there is another wrong max (the one that does take in a dim for reduction) that keeps getting matched first.

Caveats!
- We do not fast path if the shapes, dtypes, device, the regular shebang for foreach are not met. We fall back to slowpath!
- MORE IMPORTANTLY, we also do not fast path for int8 and int16 and bool, but that's really a skill issue on my end as I've hardcoded -INFINITY into the CUDA kernels, and -INFINITY is not defined for small ints. It'd be nice to know how to do this properly, but that work can also come later. 
- This does NOT support empty Tensors in the list, because the original max op also does not support empty Tensors. ~I think this should be allowed though, and this PR may come later.~ I understand why this is not allowed.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #127412
* #127303
* __->__ #127187

janeyx99 committed:
Add _foreach_max

[ghstack-poisoned]

vadimkantorov commented:
Is it more like `_foreach_amax`?

janeyx99 committed:
Update on "Add _foreach_max"

This PR adds _foreach_max support, the second reduction foreach op we have :D 

I did have to change the autogen slightly for foreach. I can promise that the existing foreach ops' derivative behavior has not changed as I've added a skip list for the harder requirement I am setting (that the arg list should match in length). I needed to add this requirement as there is another wrong max (the one that does take in a dim for reduction) that keeps getting matched first.

Caveats!
- We do not fast path if the shapes, dtypes, device, the regular shebang for foreach are not met. We fall back to slowpath!
- MORE IMPORTANTLY, we also do not fast path for int8 and int16, but that's really a skill issue on my end as I've hardcoded -INFINITY into the CUDA kernels, and -INFINITY is not defined for small ints. It'd be nice to know how to do this properly, but that work can also come later. 
- This does NOT support empty Tensors in the list, because the original max op also does not support empty Tensors. I think this should be allowed though, and this PR may come later.




[ghstack-poisoned]

janeyx99 commented:
> Is it more like `_foreach_amax`?

no, this is simpler—> there’s no dim argument

vadimkantorov commented:
> no, this is simpler—> there’s no dim argument

I see. I was thinking that the conception for `amax` variant is now to be preferred for all cases where elements' indices are not needed

crcrpar commented on aten/src/ATen/native/cuda/ForeachReduceOp.cu:
```diff
@@ -44,6 +45,191 @@ struct TensorListAddresses {
+    T vals[kILP];
+    T r_x[kILP];
+    for (int64_t i = 0; i < kILP; i++) {
+      vals[i] = T(-INFINITY);
+      r_x[i] = T(-INFINITY);
```
could be replaced with `std::numeric_limits<T>::min()`

janeyx99 replied:
experimented and used `lowest` in https://github.com/pytorch/pytorch/pull/127303

crcrpar commented on aten/src/ATen/native/cuda/block_reduce.cuh:
```diff
@@ -103,7 +103,7 @@ __inline__ __device__ T BlockReduceMax(T val, T* shared) {
   }
   __syncthreads();
-  val = (tid < B::Warps()) ? shared[lid] : T(0);
+  val = (tid < B::Warps()) ? shared[lid] : T(-INFINITY);
```
ditto (numeric limits)

janeyx99 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -11119,6 +11119,14 @@
+  device_check: NoCheck   # foreach kernels fall back to slow path when tensor are on different devices
+  variants: function
+  dispatch:
+    CPU: foreach_tensor_max_slow
```
Wait, sorry, what is the question?

albanD replied:
per offline discussion: change this key to CompositeExplicitAutograd.
Since other foreach ops can be upgraded as well, you can do that in a follow up.

janeyx99 replied:
Okay!

janeyx99 commented on aten/src/ATen/native/cuda/ForeachReduceOp.cu:
```diff
@@ -44,6 +45,191 @@ struct TensorListAddresses {
+    T vals[kILP];
+    T r_x[kILP];
+    for (int64_t i = 0; i < kILP; i++) {
+      vals[i] = T(-INFINITY);
```
Hm..I have this in the PR stacked above: https://github.com/pytorch/pytorch/pull/127303

I could merge the two changes in 1, but had opened a new PR to not mess with the CI of this one haha

janeyx99 commented:
@pytorchbot merge

janeyx99 committed:
Update on "Add _foreach_max"

This PR adds _foreach_max support, the second reduction foreach op we have :D 

I did have to change the autogen slightly for foreach. I can promise that the existing foreach ops' derivative behavior has not changed as I've added a skip list for the harder requirement I am setting (that the arg list should match in length). I needed to add this requirement as there is another wrong max (the one that does take in a dim for reduction) that keeps getting matched first.

Caveats!
- We do not fast path if the shapes, dtypes, device, the regular shebang for foreach are not met. We fall back to slowpath!
- MORE IMPORTANTLY, we also do not fast path for int8 and int16 and bool, but that's really a skill issue on my end as I've hardcoded -INFINITY into the CUDA kernels, and -INFINITY is not defined for small ints. It'd be nice to know how to do this properly, but that work can also come later. 
- This does NOT support empty Tensors in the list, because the original max op also does not support empty Tensors. ~I think this should be allowed though, and this PR may come later.~ I understand why this is not allowed.




[ghstack-poisoned]
